File: clases/20_sesion_de_problemas/codigo_morse/codigo_morse.cpp
#include <cctype>
#include <iostream>
#include <map>
#include <utility>
using namespace std;

bool EstaEnMorse(string texto) {
    for (char c : texto) {
        if (c == ' ') {
            continue;
        }
        if (isalpha(c)) {
            return false;
        }
    }
    return true;
}

int main(void) {
    map<char, string> letra_a_morse;
    letra_a_morse['A'] = ".-";
    letra_a_morse['B'] = "-...";
    letra_a_morse['C'] = "-.-.";
    letra_a_morse['D'] = "-..";
    letra_a_morse['E'] = ".";
    letra_a_morse['F'] = "..-.";
    letra_a_morse['G'] = "--.";
    letra_a_morse['H'] = "....";
    letra_a_morse['I'] = "..";
    letra_a_morse['J'] = ".---";
    letra_a_morse['K'] = "-.-";
    letra_a_morse['L'] = ".-..";
    letra_a_morse['M'] = "--";
    letra_a_morse['N'] = "-.";
    letra_a_morse['O'] = "---";
    letra_a_morse['P'] = ".--.";
    letra_a_morse['Q'] = "--.-";
    letra_a_morse['R'] = ".-.";
    letra_a_morse['S'] = "...";
    letra_a_morse['T'] = "-";
    letra_a_morse['U'] = "..-";
    letra_a_morse['V'] = "...-";
    letra_a_morse['W'] = ".--";
    letra_a_morse['X'] = "-..-";
    letra_a_morse['Y'] = "-.--";
    letra_a_morse['Z'] = "--..";

    map<string, char> morse_a_letra;
    for (pair<char, string> p : letra_a_morse) {
        char letra = p.first;
        string codigo = p.second;
        morse_a_letra[codigo] = letra;
    }

    cout << "Traductor de código Morse" << endl;
    while (true) {
        cout << "> ";
        string entrada;
        getline(cin, entrada);

        if (EstaEnMorse(entrada)) {
            for (int i = 0; i < entrada.size(); i++) {
                if (entrada[i] == '.' || entrada[i] == '-') {
                    int j = i;
                    while (j < entrada.size() && (entrada[j] == '.' || entrada[j] == '-')) {
                        j++;
                    }
                    string codigo = entrada.substr(i, j - i);
                    cout << morse_a_letra[codigo];
                    i = j;
                }
            }
        } else {
            for (char c : entrada) {
                if (c == ' ') {
                    continue;
                }
                cout << letra_a_morse[c] << ' ';
            }
        }
        cout << endl;
    }
    return 0;
}

File: tareas/2/24190108/Lsita 20/Codigo Morse.py
def esta_en_morse(texto):
    for c in texto:
        if c == ' ':
            continue
        if c.isalpha():
            return False
    return True

letra_a_morse = {
    'A': ".-", 'B': "-...", 'C': "-.-.", 'D': "-..", 'E': ".", 'F': "..-.",
    'G': "--.", 'H': "....", 'I': "..", 'J': ".---", 'K': "-.-", 'L': ".-..",
    'M': "--", 'N': "-.", 'O': "---", 'P': ".--.", 'Q': "--.-", 'R': ".-.",
    'S': "...", 'T': "-", 'U': "..-", 'V': "...-", 'W': ".--", 'X': "-..-",
    'Y': "-.--", 'Z': "--.."
}

morse_a_letra = {codigo: letra for letra, codigo in letra_a_morse.items()}

print("Traductor de código Morse")
while True:
    entrada = input("> ")

    if esta_en_morse(entrada):
        i = 0
        while i < len(entrada):
            if entrada[i] == '.' or entrada[i] == '-':
                j = i
                while j < len(entrada) and (entrada[j] == '.' or entrada[j] == '-'):
                    j += 1
                codigo = entrada[i:j]
                print(morse_a_letra.get(codigo, ''), end='')
                i = j
            else:
                i += 1
    else:
        for c in entrada:
            if c == ' ':
                continue
            print(letra_a_morse.get(c.upper(), ''), end=' ')
    print()